x/axelarnet/message_handler.go

```diff
@@ -0,0 +1,271 @@
+		Recipient:   m.Recipient,
+		PayloadHash: m.PayloadHash,
+		Payload:     msg.Payload,
+	})
+
+	return n.SetNewMessage(ctx, m)
+}
+
+func handleMessageWithToken(ctx sdk.Context, n types.Nexus, b types.BankKeeper, sourceAddress nexus.CrossChainAddress, msg Message, asset keeper.Coin) error {
+	destChain := funcs.MustOk(n.GetChain(ctx, nexus.ChainName(msg.DestinationChain)))
+
+	if err := asset.Lock(b, types.MessageSender); err != nil {
+		return err
+	}
+
+	recipient := nexus.CrossChainAddress{Chain: destChain, Address: msg.DestinationAddress}
+	m := nexus.NewGeneralMessage(
+		n.GenerateMessageID(ctx),
+		sourceAddress,
+		recipient,
+		crypto.Keccak256Hash(msg.Payload).Bytes(),
+		nexus.Sent,
+		&asset.Coin,
+	)
+
+	events.Emit(ctx, &types.ContractCallWithTokenSubmitted{
+		MessageID:   m.ID,
+		Sender:      m.Sender,
+		Recipient:   m.Recipient,
+		PayloadHash: m.PayloadHash,
+		Payload:     msg.Payload,
+		Asset:       asset.Coin,
+	})
+
+	return n.SetNewMessage(ctx, m)
+}
+
+func handleTokenSent(ctx sdk.Context, n types.Nexus, b types.BankKeeper, sourceAddress nexus.CrossChainAddress, msg Message, asset keeper.Coin) error {
+	destChain := funcs.MustOk(n.GetChain(ctx, nexus.ChainName(msg.DestinationChain)))
+	crossChainAddr := nexus.CrossChainAddress{Chain: destChain, Address: msg.DestinationAddress}
+
+	if err := asset.Lock(b, types.MessageSender); err != nil {
+		return err
+	}
+
+	transferID, err := n.EnqueueTransfer(ctx, sourceAddress.Chain, crossChainAddr, asset.Coin)
+	if err != nil {
+		return err
+	}
+
+	events.Emit(ctx, &types.TokenSent{
+		TransferID: transferID,
+		Sender:     sourceAddress,
+		Recipient:  crossChainAddr,
+		Asset:      asset.Coin,
+	})
+
+	return nil
+
+}
+
+// toICS20Packet unmarchal packet to ICS20 token packet
+func toICS20Packet(packet ibcexported.PacketI) (ibctransfertypes.FungibleTokenPacketData, error) {
+	var data ibctransfertypes.FungibleTokenPacketData
+	if err := types.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return data, sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "cannot unmarshal ICS-20 transfer packet data")
+	}
+
+	if err := data.ValidateBasic(); err != nil {
+		return data, err
+	}
+
+	return data, nil
+}
+
+// extractTokenFromPacketData get normalized token from ICS20 packet
+// panic if unable to unmarshal packet data
+func extractTokenFromPacketData(ctx sdk.Context, ibcK keeper.IBCKeeper, n types.Nexus, packet ibcexported.PacketI) (keeper.Coin, error) {
+	data := funcs.Must(toICS20Packet(packet))
+
+	// parse the transfer amount
+	amount := funcs.MustOk(sdk.NewIntFromString(data.Amount))
+
+	var denom string
+	if ibctransfertypes.ReceiverChainIsSource(packet.GetSourcePort(), packet.GetSourceChannel(), data.Denom) {
+		// sender chain is not the source, un-escrow token
+
+		// remove prefix added by sender chain
+		voucherPrefix := ibctransfertypes.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
+		unprefixedDenom := data.Denom[len(voucherPrefix):]
```

Comment:
```suggestion
		// remove prefix added by sender chain
		icsPrefix := ibctransfertypes.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
		unprefixedDenom := data.Denom[len(icsPrefix):]
```
maybe? their terminology doesn't make sense lol